content/blog/2018/appswitch/appswitch.md

```diff
@@ -0,0 +1,81 @@
+
+## Dependency ordering with AppSwitch
+
+[AppSwitch](http://appswitch.io) sits on the control path of the application’s network interactions and knows precisely when a service becomes a client by making the `connect` call and when a particular service becomes ready to accept connections by making the `listen` call.  Instead of coarsely sequencing entire services as per a dependency graph, it is conceivable that the `connect` and `listen` calls made by those services can be lined up at a finer granularity.  That would effectively solve the multilevel dependency problem and speedup application startup.
+
+But that still requires a dependency graph.  A number of products and tools exist to help with discovering service dependencies.  But they are typically based on passive monitoring of network traffic and cannot provide the information beforehand for any arbitrary application.  Network level obfuscation due to encryption and tunneling also makes them unreliable.  The burden of discovering and specifying the dependencies ultimately falls to the developer or the operator of the application.  As it is, even consistency checking a dependency specification is itself quite complex and any way to avoid requiring a dependency graph would be most desirable.
+
+The point of a dependency graph is to know which clients depend on a particular service so that those clients can then be made to wait for the respective service to become live.  But does it really matter which specific clients?  Ultimately one tautology that always holds is that all clients of a service have an implicit dependency on the service.  That’s what AppSwitch leverages to get around the requirement.  In fact, AppSwitch sidesteps dependency ordering altogether.  All services of the application can be co-scheduled without regard to any startup order.  Interdependencies among them automatically work themselves out at the granularity of individual requests and responses, resulting in fast application startups.
+
+### AppSwitch model and constructs
+
+Now that we have a conceptual understanding of AppSwitch’s high-level approach, let’s look at the constructs involved.  But first a quick summary of the usage model is in order.  Even though it is written for a different context, reviewing my earlier [blog](https://istio.io/blog/2018/delayering-istio/delayering-istio) on this topic would be useful as well.  For completeness, let me also note AppSwitch doesn’t bother with non-network dependencies.  For example it may be possible for two services to interact using IPC mechanisms or through the shared file system.  Processes with deep ties like that are typically part of the same service anyway and don’t require framework’s intervention for ordering.
```

Comment:
Here and elsewhere, use relative URLs into istio.io. That is, instead of https://istio.io/... just use /...